### PR TITLE
[Core] Vectorized search python interface

### PR DESCRIPTION
**Pull request by RiccardoRossi**
**📝 Description**

Adds a function VectorizedFind which finds the location of many points at once on the top of a tet (or triangular) mesh

**🆕 Changelog**

- [Adding VectorizedFind](https://github.com/KratosMultiphysics/Kratos/pull/12745/commits/1e95635ceb2def11a013c3410c50ae14a1acf7e6)
- [Adding test](https://github.com/KratosMultiphysics/Kratos/pull/12745/commits/a8376ede0a1cfb990fc52f8ef01265f1b0db8399)
- [Parallelizing, and adding tests](https://github.com/KratosMultiphysics/Kratos/pull/12745/commits/ad7a9ee546005d2081887e1889bb7604eec6e32f)
- [Improving implementation](https://github.com/KratosMultiphysics/Kratos/pull/12745/commits/96d2b5a57d844e3cb1bb03779c634191645904c7)

### CODE REVIEW

File: kratos/python/add_geometrical_utilities_to_python.cpp [hunk 1]
```diff
@@ -17,6 +17,7 @@
 
 // Project includes
 #include "includes/define_python.h"
+#include <pybind11/numpy.h>
```
Comment (loumalouomega) [inline kratos/python/add_geometrical_utilities_to_python.cpp]: This is an external dependency

File: applications/FluidDynamicsApplication/tests/test_vectorized_interpolation.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        ##check for the result of interpolation
+        err_norm = np.linalg.norm(xyz - interpolated_data)
+        self.assertAlmostEqual(err_norm, 0.0)
+
+
+
+    def tearDown(self):
+        pass
+
+
```
Comment (loumalouomega) [inline applications/FluidDynamicsApplication/tests/test_vectorized_interpolation.py]:
```suggestion
```
Left over

File: kratos/tests/test_vectorized_interpolation.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+        ##check for the result of interpolation
+        err_norm = np.linalg.norm(xyz - interpolated_data)
+        self.assertAlmostEqual(err_norm, 0.0)
+
+
+
+    def tearDown(self):
+        pass
+
```
Comment (loumalouomega) [inline kratos/tests/test_vectorized_interpolation.py]:
```suggestion
```

Leftover

**Review by loumalouomega — COMMENTED** [review]
Minor comments

**Review by loumalouomega — DISMISSED** [review]

File: kratos/python/add_geometrical_utilities_to_python.cpp (hunk 1 above)
Comment (RiccardoRossi) [reply kratos/python/add_geometrical_utilities_to_python.cpp]: no, it is only the interface to pybind. We already include it with Kratos together with the pybind code

Comment (loumalouomega) [reply kratos/python/add_geometrical_utilities_to_python.cpp]: I mean it is not a "Project includes", but "External includes"

**Comment by RiccardoRossi** [comment]
ok, i see ... well let me see first if it compiles and i'll change it (however please note that define_python.h contains the pybind11 headers)

**Review by RiccardoRossi — COMMENTED** [review]
Mmm @roigcarlo the fails are strange

.they seem to be completely unrelated...can you take a quick look?

**Comment by RiccardoRossi** [comment]
> Mmm @roigcarlo the fails are strange
> 
> .they seem to be completely unrelated...can you take a quick look?

@roigcarlo i was obviously very wrong. I don't know what output i looked into, the error was obviously mine. let's see if i fixed it

File: kratos/python/add_geometrical_utilities_to_python.cpp [hunk 4]
```diff
@@ -298,6 +342,7 @@ void AddGeometricalUtilitiesToPython(pybind11::module &m)
             const bool is_found = rSelf.FindPointOnMeshSimplified(rCoords,N,p_elem);
             return std::tuple<bool,Vector,Element::Pointer>{is_found,N,p_elem};
         })
+        .def("VectorizedFind", &VectorizedFindHelper<3>)
```
Comment (philbucher) [inline kratos/python/add_geometrical_utilities_to_python.cpp:345]:
```suggestion
        .def("ParallelFind", &ParallelFindHelper<3>)
```

Makes it more clear IMO

Comment (RiccardoRossi) [reply kratos/python/add_geometrical_utilities_to_python.cpp:345]: Hi @philbucher this is about allowing a vectorized use from python (not necessarily parallel) 
so in my opinion i would keep the name as it is if it is not a blocker to you

**Comment by RiccardoRossi** [comment]
@loumalouomega can you please reapprove?

**Review by loumalouomega — APPROVED** [review]

**Review by RiccardoRossi — COMMENTED** [review]
sorry @philbucher 

i thought i posted my comment. If you see it as blocker i'll generate another PR with this name change

File: kratos/python/add_geometrical_utilities_to_python.cpp (hunk 4 above)
Comment (philbucher) [reply kratos/python/add_geometrical_utilities_to_python.cpp:345]: I was confused why you used this name. you are using indexpartition (which uses omp aka shared memory parallel) and not vectorization in the simd sense

Do you have plans to change the backend?

**Comment by RiccardoRossi** [comment]
In essence the goal of this PR is to input a matrix containing XYZs coordinates of points and to retrieve **pure numpy** data structures. 

The good thing of this is that you can then do everythign efficiently on the python side on the basis of the python library. 

take a look at the function "testVectorizedInterpolation(self):" in the test, i hope it is self explicative (everything is in python, but it should be fairly fast)

**Pull request merged**